### ISSUE TITLE
How to validate non-required data

### ISSUE THREAD

**Issue by sldhana** [jamesturk/validictory#105]
I have a schema like this.  The `secondary_email` should only be validated if it contains anything other than an blank value.  But with this schema, the pattern is always being used to validate a blank value and it fails validation. 

```
data = {"testing": "", 'email': 'yoo@y.com', 'secondary_email': ''}
schema = {
    "properties": {
        "email": {
            "required": True,
            'pattern': "^[a-zA-Z0-9'\-_\+]+(?:[\.][a-zA-Z0-9'\-_\+]+)*@([a-zA-Z0-9]([-]?[a-zA-Z0-9])*\.)*[a-zA-Z0-9]([-]?[a-zA-Z0-9]){0,61}[a-zA-Z0-9]{0,1}\.[a-zA-Z]{2,6}(\.[a-zA-Z]{2})?$"
        },
        "secondary_email": {
            'type': ['string', 'null'],
            'blank': True,
            'required': False,
            'pattern': "^[a-zA-Z0-9'\-_\+]+(?:[\.][a-zA-Z0-9'\-_\+]+)*@([a-zA-Z0-9]([-]?[a-zA-Z0-9])*\.)*[a-zA-Z0-9]([-]?[a-zA-Z0-9]){0,61}[a-zA-Z0-9]{0,1}\.[a-zA-Z]{2,6}(\.[a-zA-Z]{2})?$"
        },
        "testing": {
            "required": False,
            "blank": True # fails
        }
    }
}
```

Next approach was to do this:

```
data = {"testing": "", 'email': 'yoo@y.com', 'secondary_email': ''}
schema = {
    "properties": {
        "email": {
            "required": True,
            'pattern': "^[a-zA-Z0-9'\-_\+]+(?:[\.][a-zA-Z0-9'\-_\+]+)*@([a-zA-Z0-9]([-]?[a-zA-Z0-9])*\.)*[a-zA-Z0-9]([-]?[a-zA-Z0-9]){0,61}[a-zA-Z0-9]{0,1}\.[a-zA-Z]{2,6}(\.[a-zA-Z]{2})?$"
        },
        "secondary_email": {
            'type': ['string', 'null'],
            'blank': True,
            'required': False,
            'pattern': "^[a-zA-Z0-9'\-_\+]+(?:[\.][a-zA-Z0-9'\-_\+]+)*@([a-zA-Z0-9]([-]?[a-zA-Z0-9])*\.)*[a-zA-Z0-9]([-]?[a-zA-Z0-9]){0,61}[a-zA-Z0-9]{0,1}\.[a-zA-Z]{2,6}(\.[a-zA-Z]{2})?$"
        },
        "testing": {
            "required": False,
            "blank": True # fails
        }
    }
}

```

I then have a custom email function that checks the `format_option` and determine whether it contains the word `optional` or not.  If optional, I then only validate if  `value` is not empty.    Although this works, this solution doesn't seem very elegant.  Am I missing something that I could do much simpler?

**Comment by jamesturk** [comment]
Hmm, if pattern is present it is validated... I see the case for making blank override this but I'm not sure if that'd break compatibility w/ JSON schema.  Is it an option to make your secondary regex accept blank strings as well?

**Comment by sldhana** [comment]
@jamesturk - seems like then I have to specify the regex differently for each optional field.  I am trying to do this with dynamic form generation. Could you explain the compatibility with JSON schema?  Is it because each key should be independent and not rely on each other?

**Comment by jamesturk** [comment]
That's fair.  I'll try and figure out which is expected behavior from JSON
schema's point of view (it's basically a question of precedence, IIRC this
might be vague in the draft spec)

**Comment by jamesturk** [comment]
closed in light of #114
